flink-libraries/flink-table/src/main/java/org/apache/flink/table/runtime/functions/aggfunctions/cardinality/HyperLogLog.java

```diff
@@ -0,0 +1,333 @@
+		DataOutput dos = new DataOutputStream(baos);
+		writeBytes(dos);
+
+		return baos.toByteArray();
+	}
+
+	private void writeBytes(DataOutput serializedByteStream) throws IOException {
+		serializedByteStream.writeInt(log2m);
+		serializedByteStream.writeInt(registerSet.size * 4);
+		for (int x : registerSet.readOnlyBits()) {
+			serializedByteStream.writeInt(x);
+		}
+	}
+
+	/**
+	 * Add all the elements of the other set to this set.
+	 * <p/>
+	 * This operation does not imply a loss of precision.
+	 *
+	 * @param other A compatible Hyperloglog instance (same log2m)
+	 * @throws Exception if other is not compatible
+	 */
+	public void addAll(HyperLogLog other) throws Exception {
+		if (this.sizeof() != other.sizeof()) {
+			throw new Exception("Cannot merge estimators of different sizes");
+		}
+
+		registerSet.merge(other.registerSet);
+	}
+
+	@Override
+	public ICardinality merge(ICardinality... estimators) throws Exception {
+		HyperLogLog merged = new HyperLogLog(log2m, new RegisterSet(this.registerSet.count));
+		merged.addAll(this);
+
+		if (estimators == null) {
+			return merged;
+		}
+
+		for (ICardinality estimator : estimators) {
+			if (!(estimator instanceof HyperLogLog)) {
+				throw new Exception("Cannot merge estimators of different class");
+			}
+			HyperLogLog hll = (HyperLogLog) estimator;
+			merged.addAll(hll);
+		}
+
+		return merged;
+	}
+
+	private Object writeReplace() {
+		return new SerializationHolder(this);
+	}
+
+	/**
+	 * This class exists to support Externalizable semantics for
+	 * HyperLogLog objects without having to expose a public
+	 * constructor, public write/read methods, or pretend final
+	 * fields aren't final.
+	 *
+	 * <p>
+	 * In short, Externalizable allows you to skip some of the more
+	 * verbose meta-data default Serializable gets you, but still
+	 * includes the class name. In that sense, there is some cost
+	 * to this holder object because it has a longer class name. I
+	 * imagine people who care about optimizing for that have their
+	 * own work-around for long class names in general, or just use
+	 * a custom serialization framework. Therefore we make no attempt
+	 * to optimize that here (eg. by raising this from an inner class
+	 * and giving it an unhelpful name).
+	 * </p>
+	 */
+	private static class SerializationHolder implements Externalizable {
+
+		HyperLogLog hyperLogLogHolder;
+
+		public SerializationHolder(HyperLogLog hyperLogLogHolder) {
+			this.hyperLogLogHolder = hyperLogLogHolder;
+		}
+
+		@Override
+		public void writeExternal(ObjectOutput out) throws IOException {
+			hyperLogLogHolder.writeBytes(out);
+		}
+
+		@Override
+		public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
+			hyperLogLogHolder = Builder.build(in);
+		}
+
+		private Object readResolve() {
+			return hyperLogLogHolder;
+		}
+	}
+
+	/**
+	 * Build a HyperLogLog instance.
+	 */
+	public static class Builder implements Serializable {
```

Comment:
Do we really need a builder when there are only one modifiable parameter? Why pay for an instance of a Builder when you can have normal static factory methods? We can argue the JIT may make it irrelevant, but it's hard to be definitive about the JIT.